metaworld/policies/mujoco/sawyer_xyz/base.py

```diff
@@ -0,0 +1,67 @@
+import abc
+import warnings
+
+import numpy as np
+
+from metaworld.policies.base import Policy
+
+
+class SawyerXYZPolicy(Policy, metaclass=abc.ABCMeta):
+    def __init__(self, pid_xyz):
+        self._pid_xyz = pid_xyz
+
+    def get_action(self, obs):
+        """Gets an action in response to an observation.
+
+        Args:
+            obs (np.ndarray): Observation which conforms to env.observation_space
+
+        Returns:
+            np.ndarray: 4 element array representing the action to take
+        """
+        xyz_curr = obs[self.obs_idx_of_effector_xyz]
+        xyz_error = self.xyz_goal(obs) - xyz_curr
+        xyz_response = self._pid_xyz.response_to(xyz_error)
+
+        if np.any(np.absolute(xyz_response) > 1.):
+            warnings.warn('PID constants may be too high. Response gets clipped at [-1, 1]')
+
+        action = np.zeros(4)
+        action[self.act_idx_of_effector_xyz] = xyz_response
+        action[self.act_idx_of_effector_grab] = self.grab_amount(obs)
+        return action
+
+    @abc.abstractmethod
+    def xyz_goal(self, obs):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def grab_amount(self, obs):
+        raise NotImplementedError
+
+    @property
+    def obs_idx_of_effector_xyz(self):
```

Comment:
i think this would be more succinct in the form of a function `extract_obs` which returns a dict or namespace of expected components (e.g. `o_d = self.extract_obs(o); assert o_d['end_effector_xyz'] is not None`), etc.

You can redefine it for each policy/environment when there are new elements.

Reply:
by "this" i mean the proprrties (obs_idx_...) and (act_idx_...)